cedar-policy-validator/src/types.rs

```diff
@@ -904,9 +1022,40 @@ impl EntityRecordKind {
         }
     }
 
+    /// Return `true` if this entity or record may have additional undeclared
+    /// attributes.
+    pub(crate) fn has_open_attributes_record(&self, schema: &ValidatorSchema) -> bool {
+        match self {
+            // Records explicitly store this information.
+            EntityRecordKind::Record {
+                open_attributes, ..
+            } => *open_attributes,
+            // We know Actions never have additional attributes. This is true
+            // because the upper bound for any two action entities is
+            // `AnyEntity`, so if we have an ActionEntity here its attributes
+            // are known precisely.
+            EntityRecordKind::ActionEntity { .. } => false,
+            // The `AnyEntity` type has no declared attributes, but it is a
+            // super type of all other entity types which may have attributes,
+            // so it clearly may have additional attributes.
+            EntityRecordKind::AnyEntity => true,
+            // An entity LUB may have additional attributes if any of the
+            // elements may have additional attributes.
+            EntityRecordKind::Entity(lub) => lub.iter().any(|e_name| {
+                schema
+                    .get_entity_type(e_name)
+                    .map(|e_type| e_type.open_attributes)
+                    // The entity type was not found in the schema, so we know
+                    // nothing about it and must assume that it may have
+                    // additional attributes.
+                    .unwrap_or(true)
+            }),
+        }
+    }
+
     pub(crate) fn get_attr(&self, schema: &ValidatorSchema, attr: &str) -> Option<AttributeType> {
         match self {
-            EntityRecordKind::Record { attrs } => attrs.get_attr(attr).cloned(),
+            EntityRecordKind::Record { attrs, .. } => attrs.get_attr(attr).cloned(),
```

Comment:
But if you have additional field types, shouldn't you possibly return `Never` for the attribute?